src/block.ts

```diff
@@ -0,0 +1,269 @@
+
+const Trie = require('merkle-patricia-tree')
+
+/**
+ * An object that represents the block
+ */
+export class Block {
+  public readonly header: BlockHeader
+  public readonly transactions: Transaction[] = []
+  public readonly uncleHeaders: BlockHeader[] = []
+  public readonly txTrie = new Trie()
+
+  private readonly _common: Common
+
+  /**
+   * Creates a new block object
+   *
+   *
+   * @param data - The block's data.
+   * @param opts - The network options for this block, and its header, uncle headers and txs.
+   */
+  constructor(
+    data: Buffer | [Buffer[], Buffer[], Buffer[]] | BlockData = {},
+    opts: NetworkOptions = {},
+  ) {
+    if (opts.common) {
+      if (opts.chain !== undefined || opts.hardfork !== undefined) {
+        throw new Error(
+          'Instantiation with both opts.common and opts.chain / opts.hardfork parameter not allowed!',
+        )
+      }
+
+      this._common = opts.common
+    } else {
+      const chain = opts.chain ? opts.chain : 'mainnet'
+      const hardfork = opts.hardfork ? opts.hardfork : null
+      this._common = new Common(chain, hardfork)
+    }
+
+    let rawTransactions
+    let rawUncleHeaders
+
+    if (Buffer.isBuffer(data)) {
+      // We do this to silence a TS error. We know that after this statement, data is
+      // a [Buffer[], Buffer[], Buffer[]]
+      const dataAsAny = rlp.decode(data) as any
+      data = dataAsAny as [Buffer[], Buffer[], Buffer[]]
+    }
+
+    if (Array.isArray(data)) {
+      // TODO: Pass the common object
+      this.header = new BlockHeader(data[0], opts)
+      rawTransactions = data[1]
+      rawUncleHeaders = data[2]
+    } else {
+      // TODO: Pass the common object
+      this.header = new BlockHeader(data.header, opts)
+      rawTransactions = data.transactions || []
+      rawUncleHeaders = data.uncleHeaders || []
+    }
+
+    // parse uncle headers
+    for (let i = 0; i < rawUncleHeaders.length; i++) {
+      this.uncleHeaders.push(new BlockHeader(rawUncleHeaders[i], opts))
+    }
+
+    // parse transactions
+    for (let i = 0; i < rawTransactions.length; i++) {
+      const tx = new Transaction(rawTransactions[i], opts)
+      this.transactions.push(tx)
+    }
+  }
+
+  get raw(): [Buffer[], Buffer[], Buffer[]] {
+    return this.serialize(false)
+  }
+
+  /**
+   * Produces a hash the RLP of the block
+   */
+  hash(): Buffer {
+    return this.header.hash()
+  }
+
+  /**
+   * Determines if this block is the genesis block
+   */
+  isGenesis(): boolean {
+    return this.header.isGenesis()
+  }
+
+  /**
+   * Turns the block into the canonical genesis block
+   */
+  setGenesisParams(): void {
+    this.header.setGenesisParams()
+  }
+
+  /**
+   * Produces a serialization of the block.
+   *
+   * @param rlpEncode - If `true`, the returned object is the RLP encoded data as seen by the
+   * Ethereum wire protocol. If `true`, a tuple with the raw data of the header, the txs and the
```

Comment:
Second `true` in the comment seems to be wrong if I read correctly.

Reply:
The second `true` was false. Thanks.